server/coordinator/procedure/manager_impl.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
+
+package procedure
+
+import (
+	"context"
+	"sync"
+
+	"github.com/CeresDB/ceresmeta/server/cluster"
+	"github.com/CeresDB/ceresmeta/server/coordinator/eventdispatch"
+	"github.com/CeresDB/ceresmeta/server/id"
+	"github.com/pkg/errors"
+	clientv3 "go.etcd.io/etcd/client/v3"
+)
+
+type ManagerImpl struct {
+	lock       sync.RWMutex
+	procedures []Procedure
+
+	storage  Storage
+	cluster  *cluster.Cluster
+	dispatch eventdispatch.Dispatch
+
+	procedureChan        chan<- Procedure
+	procedureIDAllocator id.Allocator
+	shardIDAllocator     id.Allocator
+}
+
+func (m *ManagerImpl) SubmitScatterTask(cxt context.Context, _ *ScatterRequest) error {
+	procedureID, err := m.procedureIDAllocator.Alloc(cxt)
+	if err != nil {
+		return errors.WithMessage(err, "alloc procedure id failed")
+	}
+	procedure := NewScatterProcedure(m.dispatch, m.cluster, procedureID, m.shardIDAllocator)
+	return m.submit(cxt, procedure)
+}
+
+func (m *ManagerImpl) SubmitTransferLeaderTask(_ context.Context, _ *TransferLeaderRequest) error {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m *ManagerImpl) SubmitMigrateTask(_ context.Context, _ *MigrateRequest) error {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m *ManagerImpl) SubmitSplitTask(_ context.Context, _ *SplitRequest) error {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m *ManagerImpl) SubmitMergeTask(_ context.Context, _ *MergeRequest) error {
+	// TODO implement me
+	panic("implement me")
+}
+
+func (m *ManagerImpl) Cancel(ctx context.Context, procedureID uint64) error {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	for _, procedure := range m.procedures {
+		if procedure.ID() == procedureID {
+			err := procedure.Cancel(ctx)
+			return errors.WithMessagef(err, "cancel procedure failed, procedureID = %d", procedureID)
+		}
+	}
+	return nil
+}
+
+func (m *ManagerImpl) ListRunningProcedure(_ context.Context) ([]*RunningProcedureInfo, error) {
+	m.lock.RLock()
+	defer m.lock.RUnlock()
+	procedureInfos := make([]*RunningProcedureInfo, 0)
+	for _, procedure := range m.procedures {
+		procedureInfos = append(procedureInfos, &RunningProcedureInfo{
+			ID:    procedure.ID(),
+			Typ:   procedure.Typ(),
+			State: procedure.State(),
+		})
+	}
+	return procedureInfos, nil
+}
+
+type RunningProcedureInfo struct {
+	ID    uint64
+	Typ   Typ
+	State State
+}
+
+func NewManagerImpl(ctx context.Context, cluster *cluster.Cluster, client *clientv3.Client, rootPath string) (Manager, error) {
+	manager := &ManagerImpl{
+		storage:              NewEtcdStorageImpl(client, cluster.GetClusterID(), rootPath),
+		cluster:              cluster,
+		dispatch:             eventdispatch.NewDispatchImpl(),
+		shardIDAllocator:     id.NewReusableAllocatorImpl(make([]uint64, 0), 0),
```

Comment:
Shouldn't the `shardIDAllocator` be passed into? It is weird to init it in a procedure manager.